rst mode improvements

robin-jarry opened the pull request:
* support multi-line inline markup
* fix comment about headings
* add support for substitution references
* add more rst example text in kitchen-sink

nightwing commented on lib/ace/mode/rst_highlight_rules.js:
```diff
@@ -171,6 +178,66 @@ var RSTHighlightRules = function() {
+     * The parser returns to the "start" state when reaching "**" */
+    "bold": [
+      {
+        token : "markup.bold",
```
It is better to use `defaultToken: "markup.bold",` instead of creating regexp, to exclude  "**"

Something like

```js
{  defaultToken: "string"},
{ token: "string.end", regex: /\*\*/, next: "start" }
```
`/\*\*/` returns to the start, and `defaultToken` assigns class to all unmatched text. 

Btw which code did you use as example, i probably need to update documentation to remove the description of old method.

nightwing commented on lib/ace/mode/rst_highlight_rules.js:
```diff
@@ -171,6 +178,66 @@ var RSTHighlightRules = function() {
+     * The parser returns to the "start" state when reaching "``" */
+    "code": [
+      {
+        token : "markup.raw",
```
it might be better to use classes which are highlighted by themes, similar to markdown mode
https://github.com/ajaxorg/ace/blob/master/lib/ace/mode/markdown_highlight_rules.js#L106
or maybe better to asciidoc mode
https://github.com/ajaxorg/ace/blob/master/lib/ace/mode/asciidoc_highlight_rules.js#L199

robin-jarry commented:
Hi @nightwing,

Following your advice, I modified some of the commits and I added a new one. Also, I added the source where I have taken the rst example.

Please tell me if there are other changes to make.

Cheers

nightwing commented on lib/ace/mode/rst_highlight_rules.js:
```diff
@@ -37,130 +37,151 @@ var TextHighlightRules = require("./text_highlight_rules").TextHighlightRules;
-        token : "markup.italic",
-        regex : "\\*(?:\\\\\\*|[^\\*])+?\\*"
+        token : tokens.italic,
+        regex : "\\*",
```
rst seems to use more complex rules for this.
`*` starts italic span only if it is preceded by space or line start and followed by non whitespace character
empty lines seem to exit italic mode as well.

so maybe we should use something like
```js
{
    token : tokens.italic,
    regex : /( |^)\*(?=[^s*])/,
    next: "italic"
}

"italic": [
    {
        token : tokens.italic,
        regex : /\*(?=\W|$)/,
        next  : "start"
    },
    {
        token : tokens.italic,
        regex : /^\s*$/,
        next  : "start"
    },
    {
        defaultToken: tokens.italic
    }
],

robin-jarry commented:
Hi @nightwing,

You're right :) I had a look at the python source code of the [official parser](http://repo.or.cz/docutils.git/blob/HEAD:/docutils/docutils/parsers/rst/states.py). This should be better now.

There are still some missing features (such as [field lists](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#field-lists)). But the tokenizer should cover most rst constructs by now.

Also, there is one (major?) limitation. If you have literal code blocks (::) inside already indented content, the highlighter will not exit the "codeblock" state until the first **zero-indented** line. The python rst parser exits the codeblock on the first line with the same level of indentation that where the block started. I don't know how to do that with ace tokenizer API.

nightwing commented:
To detect indentation level it is possible to use something like https://github.com/ajaxorg/ace/blob/master/lib/ace/mode/jade_highlight_rules.js#L155-L171

Do you think a rule to exit bold state when new paragraph is started is not needed?
```js
    {
        token : tokens.italic,
        regex : /^\s*$/,
        next  : "start"
    },
```
At least on github wiki, styles like  bold and italic are not continued to new paragraphs.

robin-jarry commented:
Ah cool, I'll have a look at the onMatch callback.

I am pretty sure that new lines are allowed in inline rst markup. At least in the official python parser implementation. I cannot check now, I'm on vacation :-)

I'll have a look at that when I get back if you want.

nightwing commented:
Merged since this pr already adds many improvements. Thank you!

Remaining issues are relatively minor and can wait until you are back from vacation :)
